### PR TITLE
[WFCORE-3187] add-user.sh tool prints unexpected characters if arrow …

### PR DESCRIPTION
**Pull request by MMarus**
…keys are pushed during interactive mode

Moved https://github.com/wildfly/wildfly-core/pull/2715 to a new rebased PR for easier review.

JBEAP: https://issues.jboss.org/browse/JBEAP-6640
WFCORE: https://issues.jboss.org/browse/WFCORE-3187

### CODE REVIEW

File: domain-management/src/main/java/org/jboss/as/domain/management/security/adduser/JavaConsole.java [hunk 1]
```diff
@@ -24,10 +24,14 @@
 
 package org.jboss.as.domain.management.security.adduser;
 
+import org.aesh.readline.Prompt;
```
Comment (darranl) [inline domain-management/src/main/java/org/jboss/as/domain/management/security/adduser/JavaConsole.java]: These imports are in the wrong location, they should be after the static imports but before the java.* imports.

File: domain-management/src/main/java/org/jboss/as/domain/management/security/adduser/ReadLineHandler.java [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+import org.aesh.readline.Readline;
+import org.aesh.readline.ReadlineBuilder;
+import org.aesh.terminal.Connection;
+import org.aesh.terminal.tty.Signal;
+
+/**
+ * Class to read line from terminal using Aesh-readline library.
+ * Created by Marek Marusic <mmarusic@redhat.com> on 8/7/17.
+ */
+public class ReadLineHandler implements Consumer<Connection> {
```
Comment (darranl) [inline domain-management/src/main/java/org/jboss/as/domain/management/security/adduser/ReadLineHandler.java]: Is this only used by classes in this package?  If so please reduce the visibility.

File: domain-management/src/test/java/org/jboss/as/domain/management/security/adduser/AeshJavaConsoleTest.java [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+import org.aesh.readline.tty.terminal.TerminalConnection;
+import org.junit.Test;
+
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.io.PipedInputStream;
+import java.io.PipedOutputStream;
+import java.nio.charset.Charset;
+
+import static org.junit.Assert.*;
```
Comment (darranl) [inline domain-management/src/test/java/org/jboss/as/domain/management/security/adduser/AeshJavaConsoleTest.java]: The static imports need to be the first set of imports,

File: core-feature-pack/src/main/resources/modules/system/layers/base/org/jboss/as/domain-add-user/main/module.xml [hunk 4]
```diff
@@ -34,6 +34,7 @@
     </resources>
 
     <dependencies>
+        <module name="org.aesh"/>
```
Comment (darranl) [inline core-feature-pack/src/main/resources/modules/system/layers/base/org/jboss/as/domain-add-user/main/module.xml]: Was the org.aesh import really needed in the end on this one?  This module is predominantly a dependency on org.jboss.as.domain-management with a main class specified.

File: domain-management/pom.xml [hunk 5]
```diff
@@ -71,6 +71,10 @@
     </dependencyManagement>
 
     <dependencies>
+        <dependency>
+            <groupId>org.aesh</groupId>
```
Comment (darranl) [inline domain-management/pom.xml]: I will let @bstansberry also comment on this dependency.

IMO I am not overly concerned as the add-user utility is destined to be deprecated and either re-written or replaced hopefully soon.

**Comment by bstansberry** [comment]
I don't want aesh as a dependency of the server runtime domain-management module or the domain-management maven module. If it has to be used we need to split this tool into a separate module. Really that should be done anyway; people wanting a slim server shouldn't have to carry a client tool.

We need to narrow our server dependency graph, not expand it.

**Comment by darranl** [comment]
Ok so this is adding a dependency on aesh which the CLI already has a dependency on, I wonder if we should talk to @jfdenise about co locating with the CLI.

For add-user 2.0 we have some other options to look at anyway but I would like to see if we can have a closer relationship with the CLI as we have some options to tie into management opts as well (if we want to take those options) 

Or another option is we move it over to wildfly-elytron-tool if we really do want a stand alone utility which also has it's own merits.

**Comment by jfdenise** [comment]
@darranl , in WF12 we are introducing a new 'security' group of commands in CLI. User management would make full sense there. CLI offers some nice features (prompting, connections, ...).
One issue I see is the fact that generic CLI script doesn't have 'user management' related options such as -u and/or -p (that I really like). This would have to be handled by wrapper scripts calling into CLI script I guess.

File: core-feature-pack/src/main/resources/modules/system/layers/base/org/jboss/as/domain-add-user/main/module.xml (hunk 4 above)
Comment (MMarus) [reply core-feature-pack/src/main/resources/modules/system/layers/base/org/jboss/as/domain-add-user/main/module.xml]: Yes, You are right. I am sorry I didn't noticed this. I fixed it along with the other suggestions from your comments. Thank You.

**Comment by ctomc** [comment]
this is ok to test

**Comment by darranl** [comment]
@bstansberry I think we can close this one out, within the next couple of WildFly releases we expect major changes this area likely involving collaboration with different tools.

**Pull request closed without merging**